Factory/SharedCatalogFactory.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Factory;
+
+use Magento\Catalog\Model\Category;
+use Magento\Catalog\Model\Product;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\Module\Manager;
+use Magento\Framework\ObjectManagerInterface;
+
+class SharedCatalogFactory
+{
+    const SHARED_CATALOG_ENABLED_CONFIG_PATH = 'btob/website_configuration/sharedcatalog_active';
+
+    private $scopeConfig;
+    private $moduleManager;
+    private $objectManager;
+
+    public function __construct(
+        ScopeConfigInterface $scopeConfig,
+        Manager $moduleManager,
+        ObjectManagerInterface $objectManager
+    ) {
+        $this->scopeConfig = $scopeConfig;
+        $this->moduleManager = $moduleManager;
+        $this->objectManager = $objectManager;
+    }
+
+    public function isSharedCatalogEnabled($storeId)
+    {
+        return $this->scopeConfig->isSetFlag(
+            self::SHARED_CATALOG_ENABLED_CONFIG_PATH,
+            \Magento\Store\Model\ScopeInterface::SCOPE_STORE,
+            $storeId
+        );
+    }
+
+    public function isSharedCatalogModuleEnabled()
+    {
+        return $this->moduleManager->isEnabled('Magento_SharedCatalog');
+    }
+
+    public function getSharedCatalogProductItemResource()
+    {
+        return $this->isSharedCatalogModuleEnabled() ?
+            $this->objectManager->create('\Magento\SharedCatalog\Model\ResourceModel\ProductItem') : false;
+    }
+
+    public function getSharedCatalogCategoryResource()
+    {
+        return $this->isSharedCatalogModuleEnabled() ?
+            $this->objectManager->create('\Magento\SharedCatalog\Model\ResourceModel\Permission') : false;
+    }
+
+    public function isProductInSharedCatalogForCustomerGroup(Product $product, $customerGroupId)
```

Comment:
Is it right that this method (and `isCategoryInSharedCatalogForCustomerGroup` as well) is called for [each product](https://github.com/algolia/algoliasearch-magento-2/pull/695/files#diff-3788f0027b1f50082fb777481e6bc216R38) during the reindex?

This could cause a big SQL overhead and not perform really well. Can it be batched and perform only 1 SQL query for a batch of products? Similar to what is done for `sales` data [here](https://github.com/algolia/algoliasearch-magento-2/blob/master/Helper/Data.php#L431) and [here](https://github.com/algolia/algoliasearch-magento-2/blob/master/Helper/Data.php#L628-L656).